### ISSUE TITLE
Can't extend Zod object contain refine function 

### ISSUE THREAD

**Issue by Abdel-Monaam-Aouini** [colinhacks/zod#2646]
![image](https://github.com/colinhacks/zod/assets/52112750/98bdd928-527c-47cd-9b6a-9bc7615566e8)
![image](https://github.com/colinhacks/zod/assets/52112750/43558e32-f9d8-46f6-af89-d6fd76f276c6)
https://codesandbox.io/s/bold-rgb-979c5f?file=/src/App.tsx

**Comment by colinhacks** [comment]
Indeed, this is a bit of a design flaw with how refinements are currently handled. I'm looking into a fix for Zod before because this would be a breaking change.

**Comment by colinhacks** [comment]
Fixed in Zod 4, the original code now works - sort of. Calling `.partial`, `.pick`, `.omit` etc will wipe away all the refinements defined on the original object. The original refinement function assumes a certain known input type. These methods change the input type, so the implementation of the refinement function may no longer be valid. So it would be unsound to include those refinements.
